### PR TITLE
CMake: support AVX2 in MinGW

### PR DESCRIPTION
**Pull request by tamird**
@yuslepukhin 

See individual commits.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -505,12 +505,12 @@ add_library(${ROCKSDB_STATIC_LIB} STATIC ${SOURCES})
...
   ${THIRDPARTY_LIBS} ${SYSTEM_LIBS})
 
-if(WIN32)
+if(MSVC)
   set_target_properties(${ROCKSDB_STATIC_LIB} PROPERTIES
     COMPILE_FLAGS "/Fd${CMAKE_CFG_INTDIR}/${ROCKSDB_STATIC_LIB}.pdb")
 endif()
 
-if(WIN32)
+if(MSVC)
```
Comment (orgads) [inline CMakeLists.txt]: This change will disable shared library build.

**Comment by orgads** [comment]
Related (less atomic, but covers more issues than this PR): https://github.com/facebook/rocksdb/pull/2052

File: CMakeLists.txt (hunk 1 above)
Comment (tamird) [inline CMakeLists.txt]: Fixed, I think. Please have another look!

**Comment by orgads** [comment]
My PR is now done. Built with MinGW 6.2.0. It's more complete than this one. Please review it.

Is there a strict policy regarding commit atomicity on this repo? Because I did it all in a single commit.

**Comment by tamird** [comment]
Updated this to include just some cleanup which was not present in your PR, @orgads.

**Comment by orgads** [comment]
It looks like you're just reverting https://github.com/facebook/rocksdb/commit/8a8c9674601ba00b47ef9ca4be8aeb8ef7395a49. Why?

**Comment by tamird** [comment]
Not at all, just simplifying it. See https://github.com/facebook/rocksdb/pull/2051/files#diff-af3b638bc2a3e6c650974192a53c7291R46

**Comment by orgads** [comment]
Oh, missed that line. Won't it break gcc? This doesn't look like a supported flag...

**Comment by tamird** [comment]
Yeah, I just fixed that so it's under MSVC.

File: CMakeLists.txt [hunk 2]
```diff
@@ -116,12 +111,10 @@ add_library(build_version OBJECT ${BUILD_VERSION_CC})
...
   ${CMAKE_CURRENT_SOURCE_DIR}/util)
 if(MSVC)
-  if (${USE_AVX2} EQUAL 1)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi /nologo  /arch:AVX2 /EHsc /GS /Gd /GR /GF /fp:precise /Zc:wchar_t /Zc:forScope /errorReport:queue")
-  else ()
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi /nologo  /EHsc /GS /Gd /GR /GF /fp:precise /Zc:wchar_t /Zc:forScope /errorReport:queue")
-  endif ()
-
+  if (NOT DEFINED AVX2 OR ${AVX2} EQUAL 1)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /arch:AVX2")
```
Comment (orgads) [inline CMakeLists.txt]: For mingw that is `-mavx2`

**Comment by tamird** [comment]
OK, done.

File: CMakeLists.txt [hunk 3]
```diff
@@ -42,11 +42,6 @@ endif()
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/modules/")
 
 if(WIN32)
```
Comment (orgads) [inline CMakeLists.txt]: Please add:

```cmake
option(WITH_AVX2 "build with AVX2" ON)
```

And adapt the usages.

File: CMakeLists.txt [hunk 4]
```diff
@@ -110,18 +105,23 @@ if (NOT WIN32)
   string(STRIP "${ROCKSDB_VERSION_MAJOR}" ROCKSDB_VERSION_MAJOR)
 endif()
 
+if(WIN32)
```
Comment (orgads) [inline CMakeLists.txt]: if(WITH_AVX2 AND WIN32)

File: CMakeLists.txt (hunk 3 above)
Comment (tamird) [inline CMakeLists.txt]: Done.

File: CMakeLists.txt (hunk 4 above)
Comment (tamird) [inline CMakeLists.txt]: Done.

**Comment by tamird** [comment]
@orgads what needs to happen for this to merge?

**Comment by orgads** [comment]
I have no idea. I'm just a (new) guest here.

**Comment by tamird** [comment]
@siying could you take a look?

**Comment by siying** [comment]
@tamird sorry for the delay. I'm merging it now.

**Pull request closed without merging**